File: eco-core/core-plugin/src/main/java/com/willfp/eco/spigot/EcoHandler.java
package com.willfp.eco.spigot;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.Handler;
import com.willfp.eco.core.config.updating.ConfigHandler;
import com.willfp.eco.core.config.wrapper.ConfigFactory;
import com.willfp.eco.core.drops.DropQueueFactory;
import com.willfp.eco.core.events.EventManager;
import com.willfp.eco.core.extensions.ExtensionLoader;
import com.willfp.eco.core.factory.MetadataValueFactory;
import com.willfp.eco.core.factory.NamespacedKeyFactory;
import com.willfp.eco.core.factory.RunnableFactory;
import com.willfp.eco.core.gui.GUIFactory;
import com.willfp.eco.core.integrations.placeholder.PlaceholderIntegration;
import com.willfp.eco.core.proxy.Cleaner;
import com.willfp.eco.core.proxy.ProxyFactory;
import com.willfp.eco.core.scheduling.Scheduler;
import com.willfp.eco.internal.EcoCleaner;
import com.willfp.eco.internal.Plugins;
import com.willfp.eco.internal.config.EcoConfigFactory;
import com.willfp.eco.internal.config.updating.EcoConfigHandler;
import com.willfp.eco.internal.drops.EcoDropQueueFactory;
import com.willfp.eco.internal.events.EcoEventManager;
import com.willfp.eco.internal.extensions.EcoExtensionLoader;
import com.willfp.eco.internal.factory.EcoMetadataValueFactory;
import com.willfp.eco.internal.factory.EcoNamespacedKeyFactory;
import com.willfp.eco.internal.factory.EcoRunnableFactory;
import com.willfp.eco.internal.gui.EcoGUIFactory;
import com.willfp.eco.internal.i18n.LocalizationUtils;
import com.willfp.eco.internal.integrations.PlaceholderIntegrationPAPI;
import com.willfp.eco.internal.logging.EcoLogger;
import com.willfp.eco.internal.proxy.EcoProxyFactory;
import com.willfp.eco.internal.scheduling.EcoScheduler;
import org.bukkit.NamespacedKey;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Logger;

public final class EcoHandler extends EcoSpigotPlugin implements Handler {
    public EcoHandler() {
        super();
    }

    private Cleaner cleaner = null;

    @Override
    public Scheduler createScheduler(@NotNull final EcoPlugin plugin) {
        return new EcoScheduler(plugin);
    }

    @Override
    public EventManager createEventManager(@NotNull final EcoPlugin plugin) {
        return new EcoEventManager(plugin);
    }

    @Override
    public NamespacedKeyFactory createNamespacedKeyFactory(@NotNull final EcoPlugin plugin) {
        return new EcoNamespacedKeyFactory(plugin);
    }

    @Override
    public MetadataValueFactory createMetadataValueFactory(@NotNull final EcoPlugin plugin) {
        return new EcoMetadataValueFactory(plugin);
    }

    @Override
    public RunnableFactory createRunnableFactory(@NotNull final EcoPlugin plugin) {
        return new EcoRunnableFactory(plugin);
    }

    @Override
    public ExtensionLoader createExtensionLoader(@NotNull final EcoPlugin plugin) {
        return new EcoExtensionLoader(plugin);
    }

    @Override
    public ConfigHandler createConfigHandler(@NotNull final EcoPlugin plugin) {
        return new EcoConfigHandler(plugin);
    }

    @Override
    public Logger createLogger(@NotNull final EcoPlugin plugin) {
        return new EcoLogger(plugin);
    }

    @Override
    public PlaceholderIntegration createPAPIIntegration(@NotNull final EcoPlugin plugin) {
        return new PlaceholderIntegrationPAPI(plugin);
    }

    @Override
    public EcoPlugin getEcoPlugin() {
        return this;
    }

    @Override
    public ConfigFactory getConfigFactory() {
        return new EcoConfigFactory();
    }

    @Override
    public DropQueueFactory getDropQueueFactory() {
        return new EcoDropQueueFactory();
    }

    @Override
    public GUIFactory getGUIFactory() {
        return new EcoGUIFactory();
    }

    @Override
    public Cleaner getCleaner() {
        if (cleaner == null) {
            cleaner = new EcoCleaner();
        }

        return cleaner;
    }

    @Override
    public ProxyFactory createProxyFactory(@NotNull final EcoPlugin plugin) {
        return new EcoProxyFactory(plugin);
    }

    @Override
    public String getLocalizedString(@NotNull final NamespacedKey key) {
        return LocalizationUtils.getStringFromKey(key).toString();
    }

    @Override
    public void addNewPlugin(@NotNull final EcoPlugin plugin) {
        Plugins.LOADED_ECO_PLUGINS.put(plugin.getName().toLowerCase(), plugin);
    }

    @Override
    public List<String> getLoadedPlugins() {
        return new ArrayList<>(Plugins.LOADED_ECO_PLUGINS.keySet());
    }

    @Override
    public EcoPlugin getPluginByName(@NotNull final String name) {
        return Plugins.LOADED_ECO_PLUGINS.get(name.toLowerCase());
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/eco/spigot/EcoHandler.kt
package com.willfp.eco.spigot

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.Handler
import com.willfp.eco.core.config.updating.ConfigHandler
import com.willfp.eco.core.config.wrapper.ConfigFactory
import com.willfp.eco.core.drops.DropQueueFactory
import com.willfp.eco.core.events.EventManager
import com.willfp.eco.core.extensions.ExtensionLoader
import com.willfp.eco.core.factory.MetadataValueFactory
import com.willfp.eco.core.factory.NamespacedKeyFactory
import com.willfp.eco.core.factory.RunnableFactory
import com.willfp.eco.core.fast.FastItemStack
import com.willfp.eco.core.gui.GUIFactory
import com.willfp.eco.core.integrations.placeholder.PlaceholderIntegration
import com.willfp.eco.core.proxy.Cleaner
import com.willfp.eco.core.proxy.ProxyFactory
import com.willfp.eco.core.scheduling.Scheduler
import com.willfp.eco.internal.EcoCleaner
import com.willfp.eco.internal.Plugins
import com.willfp.eco.internal.config.EcoConfigFactory
import com.willfp.eco.internal.config.updating.EcoConfigHandler
import com.willfp.eco.internal.drops.EcoDropQueueFactory
import com.willfp.eco.internal.events.EcoEventManager
import com.willfp.eco.internal.extensions.EcoExtensionLoader
import com.willfp.eco.internal.factory.EcoMetadataValueFactory
import com.willfp.eco.internal.factory.EcoNamespacedKeyFactory
import com.willfp.eco.internal.factory.EcoRunnableFactory
import com.willfp.eco.internal.gui.EcoGUIFactory
import com.willfp.eco.internal.integrations.PlaceholderIntegrationPAPI
import com.willfp.eco.internal.logging.EcoLogger
import com.willfp.eco.internal.proxy.EcoProxyFactory
import com.willfp.eco.internal.scheduling.EcoScheduler
import com.willfp.eco.proxy.FastItemStackFactoryProxy
import org.bstats.bukkit.Metrics
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.inventory.ItemStack
import java.io.File
import java.util.logging.Logger

@Suppress("UNUSED")
class EcoHandler : EcoSpigotPlugin(), Handler {
    private val cleaner = EcoCleaner()

    override fun createScheduler(plugin: EcoPlugin): Scheduler {
        return EcoScheduler(plugin)
    }

    override fun createEventManager(plugin: EcoPlugin): EventManager {
        return EcoEventManager(plugin)
    }

    override fun createNamespacedKeyFactory(plugin: EcoPlugin): NamespacedKeyFactory {
        return EcoNamespacedKeyFactory(plugin)
    }

    override fun createMetadataValueFactory(plugin: EcoPlugin): MetadataValueFactory {
        return EcoMetadataValueFactory(plugin)
    }

    override fun createRunnableFactory(plugin: EcoPlugin): RunnableFactory {
        return EcoRunnableFactory(plugin)
    }

    override fun createExtensionLoader(plugin: EcoPlugin): ExtensionLoader {
        return EcoExtensionLoader(plugin)
    }

    override fun createConfigHandler(plugin: EcoPlugin): ConfigHandler {
        return EcoConfigHandler(plugin)
    }

    override fun createLogger(plugin: EcoPlugin): Logger {
        return EcoLogger(plugin)
    }

    override fun createPAPIIntegration(plugin: EcoPlugin): PlaceholderIntegration {
        return PlaceholderIntegrationPAPI(plugin)
    }

    override fun getEcoPlugin(): EcoPlugin {
        return this
    }

    override fun getConfigFactory(): ConfigFactory {
        return EcoConfigFactory()
    }

    override fun getDropQueueFactory(): DropQueueFactory {
        return EcoDropQueueFactory()
    }

    override fun getGUIFactory(): GUIFactory {
        return EcoGUIFactory()
    }

    override fun getCleaner(): Cleaner {
        return cleaner
    }

    override fun createProxyFactory(plugin: EcoPlugin): ProxyFactory {
        return EcoProxyFactory(plugin)
    }

    override fun addNewPlugin(plugin: EcoPlugin) {
        Plugins.LOADED_ECO_PLUGINS[plugin.name.lowercase()] = plugin
    }

    override fun getLoadedPlugins(): List<String> {
        return ArrayList(Plugins.LOADED_ECO_PLUGINS.keys)
    }

    override fun getPluginByName(name: String): EcoPlugin? {
        return Plugins.LOADED_ECO_PLUGINS[name.lowercase()]
    }

    override fun createFastItemStack(itemStack: ItemStack): FastItemStack {
        return getProxy(FastItemStackFactoryProxy::class.java).create(itemStack)
    }

    override fun registerBStats(plugin: EcoPlugin) {
        val bStatsFolder = File(plugin.dataFolder.parentFile, "bStats")
        val configFile = File(bStatsFolder, "config.yml")
        val config = YamlConfiguration.loadConfiguration(configFile)

        if (config.isSet("serverUuid")) {
            config.set("enabled", this.ecoPlugin.configYml.getBool("enable-bstats"))
            config.save(configFile)
        }

        Metrics(plugin, plugin.bStatsId)
    }
}